pkg/custom-provider/series_registry.go

```diff
@@ -68,29 +65,73 @@ type basicSeriesRegistry struct {
 	mu sync.RWMutex
 
 	// info maps metric info to information about the corresponding series
-	info map[provider.CustomMetricInfo]seriesInfo
+	info         map[provider.CustomMetricInfo]seriesInfo
+	externalInfo map[string]seriesInfo
 	// metrics is the list of all known metrics
 	metrics []provider.CustomMetricInfo
 
 	mapper apimeta.RESTMapper
+
+	metricLister MetricListerWithNotification
 }
 
-func (r *basicSeriesRegistry) SetSeries(newSeriesSlices [][]prom.Series, namers []MetricNamer) error {
-	if len(newSeriesSlices) != len(namers) {
-		return fmt.Errorf("need one set of series per namer")
+//NewBasicSeriesRegistry creates a SeriesRegistry driven by the data from the provided MetricLister.
+func NewBasicSeriesRegistry(lister MetricListerWithNotification, mapper apimeta.RESTMapper) SeriesRegistry {
+	var registry = basicSeriesRegistry{
+		mapper:       mapper,
+		metricLister: lister,
 	}
 
+	lister.AddNotificationReceiver(registry.onNewDataAvailable)
+
+	return &registry
+}
+
+func (r *basicSeriesRegistry) filterMetrics(result metricUpdateResult) metricUpdateResult {
+	namers := make([]MetricNamer, 0)
+	series := make([][]prom.Series, 0)
+
+	targetType := config.Custom
+
+	for i, namer := range result.namers {
+		if namer.MetricType() == targetType {
+			namers = append(namers, namer)
+			series = append(series, result.series[i])
+		}
+	}
+
+	return metricUpdateResult{
+		namers: namers,
+		series: series,
+	}
+}
+
+func (r *basicSeriesRegistry) onNewDataAvailable(result metricUpdateResult) {
```

Comment:
in general, don't name things by how they're used, name them by what they do.  For example:

- don't say `onClick`, say `submitWorkflow`
- don't say `onNewDataAvailable`, say `filterAndStoreMetrics`